File: src/python/graph/bfs/tests/shortest_path_search_tests.py
import sys
import pytest

from shortest_path_search import shortest_path


graph = {
    # first subgraph
    "ROOT": ["B", "C", "D"],
    "B": ["E"],
    "C": ["F", "G"],
    "F": ["H", "I"],

    # second subgraph
    "J": ["K", "L", "M"],
    "L": ["N", "O"]
}


def test_successful_search():
    path = shortest_path(graph, root="ROOT", target="E")
    assert path == ["ROOT", "B", "E"]


def test_empty_path_if_invalid_root():
    path = shortest_path(graph, root="invalid_vertex", target="E")
    assert path == None


def test_empty_path_for_same_root_and_target():
    path = shortest_path(graph, root="ROOT", target="ROOT")
    assert path == []


def test_empty_path_for_dicsonnected_vertices():
    path = shortest_path(graph, root="ROOT", target="J")
    assert path == []


def test_empty_path_if_invalid_target():
    path = shortest_path(graph, root="ROOT", target="invalid_vertex")
    assert path == []


if __name__ == "__main__":
    sys.exit(pytest.main([__file__]))

File: src/swift/graph/bfs/tests/ShortestPathSearchTests.swift
import XCTest


class ShortestPathSearchTests: XCTestCase {

    let graph = [
        // first subgraph
        "A": ["B", "C", "D"],
        "B": ["E"],
        "C": ["F", "G"],
        "F": ["H", "E"],

        // second subgraph
        "J": ["K", "L", "M"],
        "L": ["N", "O"],   
    ]


    func test_successful_search() {
        let path = shortestPath(in: graph, from: "A", to: "E") 
        XCTAssertEqual(path, ["A", "B", "E"])
    }

    func test_nil_if_invalid_root() {
        let path = shortestPath(in: graph, from: "InvalidVertex", to: "E") 
        XCTAssertNil(path)
    }

    func test_empty_path_for_same_root_and_target() {
        let path = shortestPath(in: graph, from: "A", to: "A") 
        XCTAssertTrue(path!.isEmpty)
    }

    func test_empty_path_for_dicsonnected_vertices() {
        let path = shortestPath(in: graph, from: "A", to: "J") 
        XCTAssertTrue(path!.isEmpty)
    }

    func test_empty_path_if_invalid_target() {
        let path = shortestPath(in: graph, from: "A", to: "InvalidVertex") 
        XCTAssertTrue(path!.isEmpty)
    }


    static var allTests = [
        ("test_successful_search", test_successful_search),
        ("test_nil_if_invalid_root", test_nil_if_invalid_root),
        ("test_empty_path_for_same_root_and_target", test_empty_path_for_same_root_and_target),
        ("test_empty_path_for_dicsonnected_vertices", test_empty_path_for_dicsonnected_vertices),
        ("test_empty_path_if_invalid_target", test_empty_path_if_invalid_target),
    ]
}


func testMain() {
    #if os(Linux)

    XCTMain([
        testCase(ShortestPathSearchTests.allTests)
    ])
    #endif
}